### ISSUE TITLE
ActionChain and Statistics enhancements

### ISSUE THREAD

**Issue by osi** [jlink/jqwik#394]
## Testing Problem

With the prior stateful testing support, a common statistics check I used was to ensure that all actions were executed at least once. This was done using something like

```java
        {
            var actionType = Statistics.label("action type");
            chain.runActions()
                   .forEach(action -> actionType.collect(action.getClass()));
            actionType.coverage(Helper::everyActionTypeAtLeastOnce);
        }
```

With the new stateful testing support, the chain doesn't expose all the possible actions. The applied transformations are exposed as strings, but that can be harder to tie back to individual actions.

## Suggested Solution

`ActionChain` _could_ automatically maintain statistics around action frequency, allowing developers to only provide coverage checks. 

I can implement this myself by having my state object be something that also tracks this.

**Comment by jlink** [comment]
The problem with recording the actual action is that the action instance is not kept around. 
Since an `ActionChain` is more or less just a slightly modified `Chain` all  `Action` instances are converted to `Transformation` objects, when calling `ActionArbitrary.addAction(anAction)`. 

As long as you're using `Action.JustTransform` or `Action.JustMutate` subclasses to implement actions, and you don't override `description()` something like the following would work:

```java
StatisticsCollector transformationType = Statistics.label("transformation type");
chain.run();
chain.transformations()
	 .forEach(transformation -> transformationType.collect(transformation));
transformationType.coverage(coverage -> {
	coverage.check(ActionType1.class.getSimpleName()).count(c -> c >= 1);
	coverage.check(ActionType2.class.getSimpleName()).count(c -> c >= 1);
	...
});
```

So, I currently don't see a way to use anything else than the transformation description for statistics. 
Maybe an extension of the coverage checking API would help, e.g.

```java
coverage.checkPattern("myTransformation<.*>").count(c -> c >= 1);
```

What do you think? I may be missing an option or two...

**Comment by osi** [comment]
That approach would work with how I have migrated my code from the older approach. I had my code structured like

```java
class DoAFoo implements Action<MyState> {

  static Arbitrary< MyState > arbitrary();
}
```

which I converted to:

```java
class DoAFoo implements Transformer<MyState> {

  static Action.Independent< MyState > action();
}
```

so the `Transformer` implementations _are_ the classes I want, and `transformation` is left un-overridden so it uses the default `toString` (which for me is the one from a `Record`)

all that to say is that the string matching approach would work. `checkPattern` would be useful as the `transformation` text will be `Record.toString()` which also includes the field values. 

thinking more, even if it is immediately turned into `Transformation` objects with the more general `Chain`, could it still be useful to have it track the usage of each transformer to ensure that the test does what is expected?

**Comment by jlink** [comment]
> thinking more, even if it is immediately turned into `Transformation` objects with the more general `Chain`, could it still be useful to have it track the usage of each transformer to ensure that the test does what is expected?

If I understand correctly you suggest something like

```java
interface ActionChain<T> {
  ...
  /**
   * Return list of all used transformer instances. 
   * Depending on how your chain is created this can be more or less useful.
   * The same transformer instances could even show up multiple times in the same list.
   *
   * <p>
   * For a chain that has not been run this list is always empty.
   * </p>
   *
   * @return list of transformer instances
   */
  List<Transformer<T>> transformers();
}
```

This could be done with little effort - given it really makes sense.

**Comment by osi** [comment]
yes! i think that would provide some amount of parity with the prior implementation.

**Comment by jlink** [comment]
> yes! i think that would provide some amount of parity with the prior implementation.

See https://github.com/jlink/jqwik/commit/ef6126174248fae3924ad7a012c4f469464d5fc1.

Released in `1.7.1-SNAPSHOT`

**Comment by jlink** [comment]
`StatisticsCoverage.checkPattern(regex)` is now available in `1.7.1-SNAPSHOT`.
It's documented here: https://jqwik.net/docs/snapshot/user-guide.html#check-coverage-of-regex-pattern

**Comment by osi** [comment]
thanks for the quick turnaround!
